### PR TITLE
Dip 125 initial marc parsing

### PR DESCRIPTION
**Pull request by JPrevost**
#### What does this PR do?

- Sets up a go project... but I will change anything we want but I needed to get started so I just made some files.
- Uses `go dep` for dependency management. I wanted to try it out and it seems fine. I can adr it if we agree it is adequate.
- allows for initial MARC parsing
- relies on a forked version of MARC21 as I needed data in a way it did not provide. We can discuss that separately, but wanted to call it out here as agreeing on that work being necessary affects some core workings on this PR.
- codifies some of the MARC parsing rules into json... that isn't really done yet but it seems like a useful way to keep in sync with the metadata team once we adjust to have it not be order dependent.

#### How can a reviewer manually see the effects of these changes?

`go run . < fixtures/test.mrc`

#### What are the relevant tickets?

- https://mitlibraries.atlassian.net/browse/DIP-125

#### Requires Full Reindexing of all Sources?
NO

#### Includes new or updated dependencies?
YES

#### Todo:
- [ ] Tests
- [ ] Documentation
- [ ] Stakeholder approval

### CODE REVIEW

**Comment by gravesm** [comment]
We may want to consider the whole rules thing. Are you imagining every format is going to use some sort of JSON rules document? It may be kind of weird for some formats to use it and others not. I think I'd be inclined to just hard code things for now and look at a more generic rules based approach later if it seems needed. We would also probably have a better idea of what we need at that point. I'm concerned we are adding complexity with a) no obvious benefit and/or b) without a good understanding of the problem.

**Comment by JPrevost** [comment]
I think the rules document significantly simplifies reading of the code as well as providing a way for the metadata team to modify / adjust the rules without code changes (in a limited sense that they could add / remove subfields of interest but not create / remove a mapping entirely). If you haven't recently, I suggest looking at the spreadsheet with the mappings and then imagine all the copy / paste that will end up happening in the code just to _not_ have a file where we have an authoritative version of the agreed upon MARC rules.

File: parsers/marc.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	return r
+}
+
+func toRecord(field []string, rule *Rules, marcRecord *marc21.Record) []string {
+	field = append(field, concatSubfields(rule.Tag, []byte(rule.Subfields), marcRecord)...)
+	return field
+}
+
+// takes a mark field tag and subfields of interest for a supplied marc record and returns them concatenated
+func concatSubfields(marcfield string, subfields []byte, marcrecord *marc21.Record) []string {
```
Comment (gravesm) [inline parsers/marc.go]: I'm not sure this is going to work. For example, if we have multiple 650 fields we will want to represent these as an array of fields in Elasticsearch, not all lumped together in a single string.

Comment (JPrevost) [reply parsers/marc.go]: This returns an array of strings though...

Comment (JPrevost) [reply parsers/marc.go]: Let me see if we have sample records that confirm this is working as intended.

Comment (JPrevost) [reply parsers/marc.go]: The `record1` fixture has 5 650 fields in the record.
```
=650  \0$aArithmetic$xJuvenile poetry.
=650  \0$aChildren's poetry, American.
=650  \1$aArithmetic$xPoetry.
=650  \1$aAmerican poetry.
=650  \1$aVisual perception.
```

Each ends up as a separate subject:


```
go run . < fixtures/record1.mrc

([]marc.record) (len=1 cap=1) {
 (marc.record) {
  identifier: (string) (len=12) "   92005291 ",
  title: (string) (len=12) "Arithmetic /",
  author: ([]string) (len=1 cap=1) {
   (string) (len=26) "Sandburg, Carl, 1878-1967."
  },
  contributor: ([]string) (len=1 cap=1) {
   (string) (len=15) "Rand, Ted, ill."
  },
  url: ([]string) <nil>,
  subject: ([]string) (len=5 cap=5) {
   (string) (len=27) "Arithmetic Juvenile poetry.",
   (string) (len=28) "Children's poetry, American.",
   (string) (len=18) "Arithmetic Poetry.",
   (string) (len=16) "American poetry.",
   (string) (len=18) "Visual perception."
  },
  isbn: ([]string) (len=1 cap=1) {
   (string) (len=12) "0152038655 :"
  }
 }
}

```

Comment (JPrevost) [reply parsers/marc.go]: I think we need a better name for this method.

Comment (gravesm) [reply parsers/marc.go]: Ah, my bad. You are correct.

**Comment by gravesm** [comment]
I'm not entirely opposed to using some sort of JSON rules format, but I think we need to sit down and have a discussion about how we imagine this working before we get too far down that road. Are changes to these rules expected to just go straight into changes in the index? The pipeline eventually is likely going to result in a record being created as a struct. We are going to have to make code changes to support data model changes. Not saying we shouldn't do it, just that we should talk about how this will work and what impacts it will have on the design.

I'll go ahead and approve as I think it's a fine starting point and we can build from there.

**Pull request merged**